src/main/java/ch/njol/skript/expressions/ExprServerIcon.java

```diff
@@ -0,0 +1,138 @@
+ */
+package ch.njol.skript.expressions;
+
+import java.lang.reflect.Field;
+import ch.njol.skript.Skript;
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.classes.Changer.ChangeMode;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.ExpressionType;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.skript.lang.util.SimpleExpression;
+import ch.njol.skript.ScriptLoader;
+import ch.njol.util.coll.CollectionUtils;
+import ch.njol.util.Kleenean;
+import com.destroystokyo.paper.event.server.PaperServerListPingEvent;
+import org.bukkit.Bukkit;
+import org.bukkit.event.Event;
+import org.bukkit.event.server.ServerListPingEvent;
+import org.bukkit.util.CachedServerIcon;
+import org.eclipse.jdt.annotation.Nullable;
+
+@Name("Server Icon")
+@Description({"Icon of the server in the server list. This returns the currently shown icon on PaperSpigot 1.12.2+, but returns the default server icon on older versions or Spigot.",
+		"Also can be set to an icon that loaded using the <a href='effects.html#EffLoadServerIcon'>load server icon</a> effect, " +
+		"or can be reset to the default icon in a <a href='events.html#server_list_ping'>server list ping</a>.",
+		"",
+		"'default server icon' returns the default server icon (server-icon.png) always and cannot be changed."})
+@Examples({"on script load:",
+		"	set {server-icons::default} to the default server icon"})
+@Since("INSERT VERSION")
+public class ExprServerIcon extends SimpleExpression<CachedServerIcon> {
+
+	static {
+		Skript.registerExpression(ExprServerIcon.class, CachedServerIcon.class, ExpressionType.PROPERTY,
+				"[the] [(1¦(default)|2¦(shown|sent))] [server] icon");
+	}
+
+	private static final boolean PAPER_EVENT_EXISTS = Skript.classExists("com.destroystokyo.paper.event.server.PaperServerListPingEvent");
+
+	private boolean isServerPingEvent, isDefault;
+
+	@SuppressWarnings({"unchecked", "null"})
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		isServerPingEvent = ScriptLoader.isCurrentEvent(ServerListPingEvent.class) || (PAPER_EVENT_EXISTS && ScriptLoader.isCurrentEvent(PaperServerListPingEvent.class));
+		isDefault = (parseResult.mark == 0 && !isServerPingEvent) || parseResult.mark == 1;
+		if (!isServerPingEvent && !isDefault) {
+			Skript.error("The 'shown' server icon expression can't be used outside of a server list ping event");
+			return false;
+		}
+		return true;
+	}
+
+	@Override
+	@Nullable
+	public CachedServerIcon[] get(Event e) {
+		CachedServerIcon icon = null;
+		if ((isServerPingEvent && !isDefault) && PAPER_EVENT_EXISTS)
+			icon = ((PaperServerListPingEvent) e).getServerIcon();
+		else
+			icon = Bukkit.getServerIcon();
+		if (!PAPER_EVENT_EXISTS) {
+			try {
+				String value = (String) icon.getClass().getDeclaredField("value").get(icon);
```

Comment:
you need to cache this field in the static block, don't do lookups at runtime if avoidable

Reply:
Hmm

Reply:
We discussed this in Discord and at the end i disabled server icon stuff on Spigot because Spigot is 🤢.